Exercise : 6.2 Load content using JSON

imvishi opened the pull request.

Commit:
Exercise : 6.2 Load content using JSON

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+
+  createTargetDivForSpecial() {
+    this.form = $(this.specialObject.formSelector);
+    const title = $('<h2 id="title"></h2>');
```
These Variables are not of much use and can be optimized using one string var for the required html

imvishi replied:
Fixed.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+    const color = $('<p id="color"></p>');
+    const img = $('<img id="image">');
+    this.div = $('<div>')
+    this.div.append(title, text, color, img);
```
methods can be chained in single statement

Ankita9818 replied:
use appropriate name for div.

Ankita9818 replied:
use appropriate name for div.

imvishi replied:
Renamed div.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  bindEvents() {
+    this.form.find('select').change((e) => {
```
The JS breaks if user selects placeholder instead of any valid option. 
Handle that case as well

imvishi replied:
Fixed.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+
+  init() {
+    this.createTargetDivForSpecial();
+    this.fetchData();
```
Currently data is loaded/cached atleast once when the JS is initialized.
It can be optimized if the data is loaded/cached on first change of the target select list.

imvishi replied:
Updated code as per your suggestion.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+
+  init() {
+    this.createTargetDivForSpecial();
+    this.fetchData();
```
Currently data is loaded/cached atleast once when the JS is initialized.
It can be optimized if the data is loaded/cached on first change of the target select list.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+    });
+  }
+
+  loadData(selectedValue) {
```
Please use an apt name, loadData method should only load data not display data on target as well.

imvishi replied:
Renamed.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+      this.specialContainer.find('#title').html(data.title);
+      this.specialContainer.find('#text').html(data.text);
+      this.specialContainer.find('#color').html(data.color);
+      this.specialContainer.find('#image').attr('src', '.' + data.image);
```
Why is . prefix added ??

imvishi replied:
. is used here to create a relative path to `/images/special.jpg` that is .. `./images/special.jpg`

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,59 @@
+
+  bindEvents() {
+    this.form.find('select').change((e) => {
+      if (e.target.selectedIndex > 0) {
```
The target div should be cleared If user selects the placeholder.

imvishi replied:
Fixed.

Ankita9818 reviewed (changes requested):
Please do not force push.

Commit:
Fixed following issues:

- Clear special container when selection placeholder.
- Use appropriate names.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,70 @@
+        if (this.specialData) {
+          this.updateSpecialContainer(e.target.value);
+        } else {
+          $.getJSON(Special.specialJsonUrl, (res) => {
```
this logic can be separated out in a method.

imvishi replied:
Separated out the logic.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  removeSubmitButton() {
+    this.form.find('input').parent().remove();
```
Add a data attr and use that to select the required form element.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  removeSubmitButton() {
+    this.form.find('input').parent().remove();
```
Add a data attr and use that to select the required form element.

imvishi replied:
Fixed.

Commit:
Fixed following issues:

- Refactored code to separate out the special selection logic.
- Used data attr to store form elements.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,84 @@
+    this.bindEvents();
+  }
+
+  initializeFormNodes() {
```
no need for this, this.form can be initialized in constructor and Please fetch submit btn using data attr.
Using ambiguous HTML tags to select elements is not a good practice.

imvishi replied:
Fixed.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  bindEvents() {
+    this.form.find('select').change((e) => {
```
Any other method to bind events?? Which one has advantage over others?

imvishi replied:
Used `.on` instead of `.select()`. Reasons to use `.on()` instead of `.select()`
1. It will also work for dynamically added elements.
2. It uses less memory because in the case of `.on()` we are adding a single handler for all elements that match our selector, but with `.select()` a separate handler gets created for every single element that matches the selector.

Commit:
Added following changes:

- Avoided ambiguous HTML tags to select elements
- Used on instead of select for event handler.

Ankita9818 commented on exercises/js/specials.js:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  bindEvents() {
+    this.form.on('change', 'select', (event) => {
```
Should have used a data attribute for *select* element selector.

Ankita9818 replied:
Should have used a data attribute for *select* element selector.
